TESTS/Level3/Pwm/Pwm.cpp

```diff
@@ -0,0 +1,135 @@
+    iin.rise(callback_pwm_rise);
+    iin.fall(callback_pwm_fall);
+
+    // Set the period
+	DEBUG_PRINTF("Duty cycle: %f, Period: %d\n", dutycycle, period);
+    pwm.period((float)period/1000);
+    pwm.write(0);
+    DEBUG_PRINTF("Waiting for %dms\n", waveform_count * period);
+
+    // Start the timer, write the duty cycle, and wait for completion
+    timer.start();
+    pwm.write(dutycycle);
+    wait_ms(waveform_count * period);
+
+    // Stop the timer, reset the IRQ for interrupts (cause it may not work on some platforms) and clear the PWM duty cycle
+    pwm.write(0);
+    iin.disable_irq(); // This is here because otherwise it fails on some platforms
+    timer.stop();
+
+    // Run post calculations for verification
+    int rc = rise_count;
+    int fc = fall_count;
+    float avgTime = (float)duty_count / waveform_count;
+    float expectedTime = (float)period * dutycycle;
+    // DEBUG_PRINTF("Expected time: %f, Avg Time: %f\n", expectedTime, avgTime);
+    // DEBUG_PRINTF("rise count = %d, fall count = %d, expected count = %d\n", rc, fc, waveform_count);
+    TEST_ASSERT_MESSAGE( std::abs(rc-fc) <= calculated_percent, "There was more than a specific variance in number of rise vs fall cycles");
+    TEST_ASSERT_MESSAGE( std::abs(waveform_count - rc) <= calculated_percent, "There was more than a specific variance in number of rise cycles seen and number expected.");
+    TEST_ASSERT_MESSAGE( std::abs(waveform_count - fc) <= calculated_percent, "There was more than a specific variance in number of fall cycles seen and number expected.");
+    // @TODO The following assert is a good check to have (comparing times) but fails on most platforms. Need to come up with a better way to do this test
+    // TEST_ASSERT_MESSAGE( std::abs(expectedTime - avgTime) <= calculated_percent,"Greater than a specific variance between expected and measured duty cycle");
+
+}
+
+// Test case that combines period and duty cycles
+void test_pwm(PinName pin, float tolerance, int iterations) {
+	DEBUG_PRINTF("Running pwm test on pin %d\n", pin);
+    TEST_ASSERT_MESSAGE(pin != NC, "Pin is NC");
+
+    // Get a random seed from the Greentea host test
+    srand(TestFramework::get_seed());
+
+    for (int i=0; i<iterations; i++) {
+        int period = rand()%91 + 10;
+        float dutycycle = (rand()%81 + 10)/100.0;
+        test_pwm_execute(pin, tolerance, 20, dutycycle, period);
+    }
+}
+
+template <int iterations>
+utest::v1::control_t test_level2_pwm(const size_t call_count) {
+	return TestFramework::test_level2_framework(TestFramework::PWM, TestFramework::CITS_PWM, &test_pwm, 0.05f, iterations);
+}
+
+Case cases[] = {
+	Case("Level 3 - PWM randomized", test_level2_pwm<10>, TestFramework::greentea_failure_handler),
+};
+
+int main() {
+	// Formulate a specification and run the tests based on the Case array
+	Specification specification(TestFramework::test_setup<200>, cases);
+    return !Harness::run(specification);
+}
```

Comment:
I don't think that level3 PWM test is working?

mbedgt: test suite 'tests-level3-pwm' ................................................................ OK in 17.67 sec
        test case: 'Level 3 - PWM randomized' ........................................................ ERROR in 0.00 sec